gpsrenda/widgets/widgets.py

```diff
@@ -0,0 +1,192 @@
+    'grade': GaugeElevationMap,
+}
+
+class SpeedWidget:
+    def __init__(self, data_source, x, y, style='hbar', units='metric'):
+        self.data_source = data_source
+        self.units = units
+        gauge_class = STYLE_TABLE[style]
+        caption = 'mph' if units == 'imperial' else 'kph'
+
+        colors = [(0, [0.2, 0.0, 0.6]),
+                  (25, [0.0, 0.6, 0.0]),
+                  (50, [0.8, 0.0, 0.0])]
+
+        if units == 'imperial':
+            for idx, (val, rgb) in enumerate(colors):
+                colors[idx] = (km_to_mi(val), rgb)
+
+        gauge = gauge_class(x, y, label="{val:.1f}", dummy_label="99.9", caption=caption,
+                            data_range=colors)
+        self.gauge = gauge
+
+    def render(self, context, t):
+        value = self.data_source.speed(t) * 60 * 60 / 1000
+        if self.units == 'imperial':
+            value = km_to_mi(value)
+        self.gauge.render(context, value)
+
+
+class CadenceWidget:
+    def __init__(self, data_source, x, y, style='hbar'):
+        self.data_source = data_source
+        gauge_class = STYLE_TABLE[style]
+        gauge = gauge_class(x, y, label="{val:.0f}", dummy_label="999", caption="rpm",
+                            data_range=[(60, [0.8, 0.0, 0.0]),
+                                        (90, [0.0, 0.6, 0.0]),
+                                        (120, [0.8, 0.0, 0.0])])
+        self.gauge = gauge
+
+    def render(self, context, t):
+        value = self.data_source.cadence(t)
+        self.gauge.render(context, value)
+
+
+class HeartRateWidget:
+    def __init__(self, data_source, x, y, style='hbar'):
+        self.data_source = data_source
+        gauge_class = STYLE_TABLE[style]
+        gauge = gauge_class(x, y, label="{val:.0f}", dummy_label="999", caption="bpm",
+                            data_range=[(100, [0.2, 0.0, 0.6]),
+                                        (200, [0.8, 0.0, 0.0])])
+        self.gauge = gauge
+
+    def render(self, context, t):
+        value = self.data_source.heart_rate(t)
+        self.gauge.render(context, value)
+
+
+class PowerWidget:
+    def __init__(self, data_source, x, y, style='hbar'):
+        self.data_source = data_source
+        gauge_class = STYLE_TABLE[style]
+        gauge = gauge_class(x, y, label="{val:.0f}", dummy_label="999", caption="W",
+                            data_range=[(0, (0, 0.6, 0)),
+                                        (300, (0.2, 0.6, 0.0)),
+                                        (800, (0.8, 0.0, 0.0))])
+        self.gauge = gauge
+
+    def render(self, context, t):
+        value = self.data_source.power(t)
+        self.gauge.render(context, value)
+
+
+class TemperatureWidget:
+    def __init__(self, data_source, x, y, style='vbar', units='metric'):
+        self.data_source = data_source
+        self.units = units
+        gauge_class = STYLE_TABLE[style]
+        suffix = '°F' if units == 'imperial' else '°C'
+
+        colors = [(5, (0.6, 0, 0)),
```

Comment:
ranges for this are also good; crit temperatures are different than copperopolis temperatures

Reply:
The temperatures at Copperopolis should always be shown in red. This seems like the correct behavior to me.

On the one hand, I agree, make everything configurable. On the other hand, these are physical quantities with pretty established "normal" ranges. 180bpm is a hard effort. 30mph is fast. 5C is cold. If these weren't creating smooth gradients and you were trying to highlight something specific like your exact FTP, I would definitely make it a config parameter.